### PR TITLE
Fixed exporting flow results when steps don't have channels

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

**Comment by nicpottier** [comment]
Unit test?

**Comment by rowanseymour** [comment]
@nicpottier added and also changed export so it uses the URN of the message (might be twitter or null) rather than always using the contact's TEL urn

File: temba/flows/models.py [hunk 1]
```diff
@@ -3303,18 +3310,15 @@ def as_org_tz(dt):
...
-                msgs.write(msg_row, 0, run_step.contact.get_urn_display(org=org, scheme=TEL_SCHEME, full=True))
-                msgs.write(msg_row, 1, run_step.contact.name)
-                arrived_on = as_org_tz(run_step.arrived_on)
-
-                msgs.write(msg_row, 2, arrived_on, date_format)
-                if run_step.step_type == RULE_SET:
-                    msgs.write(msg_row, 3, "IN")
-                else:
-                    msgs.write(msg_row, 3, "OUT")
+                urn_display = msg.contact_urn.get_display(org=org, full=True) if msg.contact_urn else ''
```
Comment (nicpottier) [inline temba/flows/models.py]: Agreed that this is needed, but can you check that this change doesn't cause us to generate an extra db hit per row exported?

File: temba/flows/models.py [hunk 2]
```diff
@@ -3175,10 +3179,25 @@ def as_org_tz(dt):
         start = time.time()
         flow_names = ", ".join([f['name'] for f in self.flows.values('name')])
 
+        urn_display_cache = {}
+
+        def get_contact_urn_display(contact):
```
Comment (rowanseymour) [inline temba/flows/models.py]: Seems to make sense to still use contact's default URN for the non-message tables as some of the rows might not have messages. This method then provides caching of fetches for those as they were being repeated across different steps

**Comment by nicpottier** [comment]
Nice!

**Pull request merged**